Add anndata analysis

ahuang11 opened the pull request:
Adds ManifoldMapAnalysis and ManifoldMapPanel and refactors the app.py into smaller modules.

Also, properly implement to_spec/from_spec by writing the uploaded h5ad to disk (maybe not a good idea) so that it can be referenced, and I do wonder how to delete the files at the end of session.

The ManifoldMapAnalysis button currently appears for any tables shown since it just depends on adata and not any specific columns.

https://github.com/user-attachments/assets/88dd73ab-9197-4e7f-a6b2-13bd282511f1

Partially addresses https://github.com/holoviz-topics/lumen-anndata/issues/4 and https://github.com/holoviz-topics/lumen-anndata/issues/6

droumis commented:
I keep getting failed attempts when asking "show me a plot of the UMAP results" right after uploading a file

<img width="1310" alt="image" src="https://github.com/user-attachments/assets/9b7e74c3-5405-498e-b7da-a541b7a5cd06" />

Can we add a prioritization to the prompt injection to use the ManifoldMap app by default to visualize any UMAP, PCA, tSNE results unless otherwise specified by the user? I think this is aligned with having a predictable experience by users while still allowing for extensibility through further specification.. e.g. if they just want a vega scatter plot instead

droumis commented on src/lumen_anndata/source.py:
```diff
@@ -44,12 +49,31 @@ class AnnDataSource(DuckDBSource):
     source_type = "anndata"
 
+    _opened = {}  # Track files: {filename: (adata_object, is_temporary)}
+
```
```suggestion
operations = param.List(default=[], doc="List of AnnDataOperation instances to apply to the data.")
```

This appears to be missing and fixes the error in [my comment ](https://github.com/holoviz-topics/lumen-anndata/pull/11#issuecomment-2963086515)

droumis reviewed (commented):
should we add hv_anndata to the deps? Also, let's re-pin lumen and have it install with [ai]

```yaml

dependencies = [
    "packaging",
    "panel >=1.5.0",
    "lumen[ai] >=0.10.1", 
    "anndata >=0.11.4",
    "pooch >=1.8.2",
    "scanpy >=1.11.1",
    "hv-anndata @ git+https://github.com/holoviz-topics/hv-anndata.git@featuremap"
]

[tool.hatch.metadata]
allow-direct-references = true
```

droumis commented on src/lumen_anndata/source.py:
```diff
@@ -576,5 +628,86 @@ def create_sql_expr_source(
-        return {}
+        filename = self._adata_store.filename
+        if filename is None:
+            filename = tempfile.mktemp(suffix=".h5ad")
```
Maybe use instead [NamedTemporaryFile()](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile), passing it the delete=False parameter.

See [deprecation warning](https://docs.python.org/3/library/tempfile.html#:~:text=Warning%20Use%20of,delete%3DFalse%20parameter%3A)

droumis replied:
```
with tempfile.NamedTemporaryFile(suffix=".h5ad", delete=False) as tmp:
    filename = tmp.name
```

etc

ahuang11 commented:
Thanks for reviewing! I'll fix those issues and let you know.

ahuang11 commented:
> I'm actually sometimes confused that it doesn't present the custom analysis buttons when it probably should (even after I explicitly ask for them).

It was a bug in https://github.com/holoviz/lumen/pull/1274

ahuang11 commented:
Can request ManifoldMapAnalysis through query
<img width="979" alt="image" src="https://github.com/user-attachments/assets/69c22c2d-aa50-49c5-b847-9712e914a141" />

droumis commented:
1. upload dataset
2. ask 'what is in this dataset' (expecting a short sentence of what an anndata file is and a summary of tables found with a short description and offer to preview)
3. error:

<details> <summary> Error Trace </summary>

```python
Traceback (most recent call last):

  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.11/site-packages/lumen/ai/agents.py", line 635, in _create_valid_sql
    sql_expr_source = source.create_sql_expr_source({expr_slug: sql_query})
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/droumis/src/lumen-anndata/src/lumen_anndata/source.py", line 615, in create_sql_expr_source
    source = super().create_sql_expr_source(tables.copy(), materialize, **params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.11/site-packages/lumen/sources/duckdb.py", line 268, in create_sql_expr_source
    self._connection.execute(table_expr)

duckdb.duckdb.ParserException: Parser Error: syntax error at or near ";"


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.11/site-packages/lumen/ai/agents.py", line 641, in _create_valid_sql
    await self._retry_output_by_line(e, messages, self._memory, sql_query, language="sql"),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.11/site-packages/lumen/ai/agents.py", line 490, in _retry_output_by_line
    return apply_changes(original_lines, result.lines_changes)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.11/site-packages/lumen/ai/utils.py", line 789, in apply_changes
    original_lines[change.line_no - 1] = change.replacement
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^

IndexError: list assignment index out of range
```

</details>



<img width="402" alt="image" src="https://github.com/user-attachments/assets/683e6ea3-6e90-47bf-9801-cf604cd67566" />

ahuang11 commented:
Wondering if you pulled and updated Lumen main branch?

droumis commented:
Agent and Assistant give contradicting results. 

1. upload anndata file
2. ask 'what type of dataset is this?'
3. Agent says it doesn't have one
4. ask 'what is in this dataset?'
5. Assistant lists tables in dataset, Agent says there are none

https://github.com/user-attachments/assets/d9889693-5e29-4ec3-9b19-fea91e53c891

ahuang11 commented on src/lumen_anndata/views.py:
```diff
@@ -0,0 +1,70 @@
+    def get_panel(self):
+        hv.Store.set_current_backend('bokeh')
+        adata = self.pipeline.source.get(self.pipeline.table, return_type="anndata")
+        if self.operation:
+            operation_callable = ComputeEmbedding if self.operation == "ComputeEmbedding" else Leiden
+            adata = operation_callable(**self.operation_kwargs).apply(adata)
```
Yeah initially I had that same thought, but I had trouble serializing it to the YAML as methods in a `param.Selector`. Maybe @philippjfr knows better.

ahuang11 commented:
Thanks for testing. I addressed the issues and feedback in the latest commit and additionally requires: https://github.com/holoviz/lumen/pull/1287

<img width="938" alt="image" src="https://github.com/user-attachments/assets/95201c1c-5fbd-41ed-8882-3e023f109b55" />

I also added a toggle to cache uploaded files to disk so it doesn't have to rewrite it every time.

> 'Compute Embeddings' -> a single, opinionated pipeline that includes PCA > neighbors > UMAP > leiden clustering.

I couldn't get this to work; kept encountering leaked semaphores... Can't tell if I'm using too large of a dataset. Can you try it?

droumis commented:
> Also to clarify, is Leiden applied before or after UMAP?

Leiden clustering is applied **after** the UMAP calculation. `sc.tl.umap(adata)` and then `sc.tl.leiden(adata, flavor="igraph", n_iterations=2)`

I think we generally want to follow the steps in the [scanpy demo tutorial](https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html#nearest-neighbor-graph-constuction-and-visualization). Also, for more contextual reference, see [chapters 9 and 10 of this book](https://www.sc-best-practices.org/preprocessing_visualization/dimensionality_reduction.html).

droumis commented:
My stream of thoughts: 

> With 'ComputeEmbeddingsAnalysis' I'm running into 'ValueError: Insufficient colors provided (256) for the categorical fields available (935)' but I didn't ask it to plot anything. And I'm confused about the flow for when UMAP results exist in the dataset but someone clicks to calculate ComputeEmbeddingsAnalysis. Also, are we correctly avoiding overwriting original data file pca/umap/clustering results? Also, as a user, I would want some indication of the pipeline being run and updates about the currently active steps.

I'm getting the feeling that this an overly complicated first attempt at an analysis pipeline given that there are a number of things that could go wrong with the extensive pipeline that is unrelated to lumen. I'm sorry for over-complicating things. I don't want this PR to go on forever.. so maybe we should pivot and simplify the initial target for a demo analysis pipeline to something much simpler... for instance.. it would useful for a user to rerun just the leiden clustering step for a set of predetermined `resolution` parameter values which then can become additional color_by fields in the manifoldmap widget.

As shown [here](https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html#manual-cell-type-annotation).
```python
for res in [0.02, 0.5, 1.0, 2.0]:
    sc.tl.leiden(
        adata, key_added=f"leiden_res_{res:4.2f}", resolution=res, flavor="igraph"
    )
```

To be clear, my revised target suggestion is now to have two 'analysis' buttons... one to launch the ManifoldMap app (with no computation) and another to compute set of Lieden results as shown above (with no visualization).

@cvaske @ahuang11 thoughts?

ahuang11 commented:
Okie I think I made all the changes requested. Be sure to update to the latest Lumen main because these were merged:
https://github.com/holoviz/lumen/pull/1292
https://github.com/holoviz/lumen/pull/1291

https://github.com/user-attachments/assets/01d85a52-e143-4f40-9e5b-9a9a83e60466

droumis commented:
Can we please set the default number of `iterations` to `2` for the Leiden analysis? I don't think the UI is really set up to have a user understand and deal with a case where it takes many iterations for convergence, so I don't think `-1` (iterate until convergence) is an appropriate default.

droumis commented:
Can you please limit the decimal places on the up/down button increment for the `resolution` widget in the config to just one (e.g. 1.2, 1.3, etc)
<img width="278" alt="image" src="https://github.com/user-attachments/assets/7fe80219-e43e-4856-91dd-277fd6412c1b" />

droumis commented:
After running the LeidenComputation.. there's no obvious output status/result of the analysis. Since I knew where to look, I could scroll all the way over in the obs table explorer to see that there was a new column.. but a message of some sort telling the user that the analysis was completed and that the result is stored in a new obs table column named <leiden_{resolution}> would be helpful, me thinks

<img width="1713" alt="image" src="https://github.com/user-attachments/assets/2f9ae15a-05d2-4f29-bdd0-948b6f9a4e2a" />

ahuang11 commented:
Addressed the primary issue of 
> WARNING:param.AnnDataSource08424: Could not refine selections for table 'umap_results_full_select': Catalog Error: Table with name obsm_X_umap does not exist!
Did you mean "obs"?

Forgot to account for materialized tables.

> Can we please set the default number of iterations to 2 for the Leiden analysis
Done

> Can you please limit the decimal places on the up/down button increment for the resolution widget in the config to just one (e.g. 1.2, 1.3, etc)

I think this is more of a Python floating point 
https://0.30000000000000004.com/
<img width="696" alt="image" src="https://github.com/user-attachments/assets/d7e08a19-757c-4ca8-9f8d-13434453b602" />

Not sure what to do about it; if we are going to do something, it'll be in Panel / Lumen to round.

> but a message of some sort telling the user that the analysis was completed and that the result is stored in a new obs table column named <leiden_{resolution}> would be helpful, me thinks

Will address in Lumen.

philippjfr reviewed (commented):
I'm okay with this but I'm fairly opposed to the introduction of `operations` on the `Source`. They are effectively a reimplementation of `Pipeline` and `Transform`, though I appreciate it's not possible out of the box because of the anndata format.
